internal/declarations/util.go

```diff
@@ -0,0 +1,243 @@
+		// A non-exported default is a nonsequitor - we usually try to remove all export modifiers
+		// from statements in ambient declarations; but a default export must retain its export modifier to be syntactically valid
+		flags ^= ast.ModifierFlagsExport
+	}
+	if flags&ast.ModifierFlagsDefault != 0 && flags&ast.ModifierFlagsAmbient != 0 {
+		flags ^= ast.ModifierFlagsAmbient // `declare` is never required alongside `default` (and would be an error if printed)
+	}
+	return flags
+}
+
+func unwrapParenthesizedExpression(o *ast.Node) *ast.Node {
+	for o.Kind == ast.KindParenthesizedExpression {
+		o = o.Expression()
+	}
+	return o
+}
+
+func isPrimitiveLiteralValue(node *ast.Node, includeBigInt bool) bool {
+	// !!! Debug.type<PrimitiveLiteral>(node);
+	switch node.Kind {
+	case ast.KindTrueKeyword,
+		ast.KindFalseKeyword,
+		ast.KindNumericLiteral,
+		ast.KindStringLiteral,
+		ast.KindNoSubstitutionTemplateLiteral:
+		return true
+	case ast.KindBigIntLiteral:
+		return includeBigInt
+	case ast.KindPrefixUnaryExpression:
+		if node.AsPrefixUnaryExpression().Operator == ast.KindMinusToken {
+			return ast.IsNumericLiteral(node.AsPrefixUnaryExpression().Operand) || (includeBigInt && ast.IsBigIntLiteral(node.AsPrefixUnaryExpression().Operand))
+		}
+		if node.AsPrefixUnaryExpression().Operator == ast.KindPlusToken {
+			return ast.IsNumericLiteral(node.AsPrefixUnaryExpression().Operand)
+		}
+		return false
+	default:
+		// !!! assertType<never>(node);
+		return false
+	}
+}
+
+func isPrivateMethodTypeParameter(host DeclarationEmitHost, node *ast.TypeParameterDeclaration) bool {
+	return node.AsNode().Parent.Kind == ast.KindMethodDeclaration && host.GetEffectiveDeclarationFlags(node.AsNode().Parent, ast.ModifierFlagsPrivate) != 0
+}
+
+// If the ExpandoFunctionDeclaration have multiple overloads, then we only need to emit properties for the last one.
+func shouldEmitFunctionProperties(input *ast.FunctionDeclaration) bool {
+	if input.Body != nil { // if it has an implementation, it must be the last one
+		return true
+	}
+
+	overloadSignatures := core.Filter(input.Symbol.Declarations, func(decl *ast.Node) bool {
+		return ast.IsFunctionDeclaration(decl)
+	})
+
+	return len(overloadSignatures) == 0 || overloadSignatures[len(overloadSignatures)-1] == input.AsNode()
+}
+
+func getFirstConstructorWithBody(node *ast.Node) *ast.Node {
+	for _, member := range node.Members() {
+		if ast.IsConstructorDeclaration(member) && ast.NodeIsPresent(member.Body()) {
+			return member
+		}
+	}
+	return nil
+}
+
+func getEffectiveBaseTypeNode(node *ast.Node) *ast.Node {
+	baseType := getClassExtendsHeritageElement(node)
+	// !!! TODO: JSDoc support
+	// if (baseType && isInJSFile(node)) {
+	//     // Prefer an @augments tag because it may have type parameters.
+	//     const tag = getJSDocAugmentsTag(node);
+	//     if (tag) {
+	//         return tag.class;
+	//     }
+	// }
+	return baseType
+}
+
+func getClassExtendsHeritageElement(node *ast.Node) *ast.Node {
```

Comment:
I recently added this function to `ast/utilities.go` in one of my completions PR.